Level 1 Classifying Product Names.ipynb

```diff
@@ -0,0 +1,377 @@
+    "def transform_name(product_name):\n",
+    "    product_name = product_name.lower()\n",
+    "    product_name = re.sub(r'[^\\w]', ' ', product_name)\n",
+    "    return product_name"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "id": "d858b216",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Processing pruned_products_1.xml\n",
+      "Processing pruned_products_2.xml\n",
+      "Processing pruned_products_3.xml\n",
+      "Processing pruned_products_4.xml\n",
+      "Processing pruned_products_5.xml\n",
+      "Processing pruned_products_6.xml\n"
+     ]
+    }
+   ],
+   "source": [
+    "pairs = []\n",
+    "for filename in os.listdir(directory):\n",
+    "        if filename.endswith(\".xml\"):\n",
+    "            print(\"Processing %s\" % filename)\n",
+    "            f = os.path.join(directory, filename)\n",
+    "            tree = ET.parse(f)\n",
+    "            root = tree.getroot()\n",
+    "            for child in root:\n",
+    "                if random.random() > 1.0:\n",
+    "                    continue\n",
+    "                # Check to make sure category name is valid\n",
+    "                if (child.find('name') is not None and child.find('name').text is not None and\n",
+    "                    child.find('categoryPath') is not None and len(child.find('categoryPath')) > 0 and\n",
+    "                    child.find('categoryPath')[len(child.find('categoryPath')) - 1][0].text is not None):\n",
+    "                      # Choose last element in categoryPath as the leaf categoryId\n",
+    "                      cat = child.find('categoryPath')[len(child.find('categoryPath')) - 1][0].text\n",
+    "                      # Replace newline chars with spaces so fastText doesn't complain\n",
+    "                      name = child.find('name').text.replace('\\n', ' ')\n",
+    "                      #output.write(\"__label__%s %s\\n\" % (cat, transform_name(name)))\n",
+    "                      pair = {\"category\": cat, \"name\": \"__label__%s %s\\n\" % (cat, transform_name(name))}\n",
+    "                      pairs.append(pair)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "id": "6467d0fd",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "df = pd.DataFrame(pairs)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "id": "b0bb0d47",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "1952"
+      ]
+     },
+     "execution_count": 8,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "df['category'].unique().size"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "id": "5cbe5e24",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# 1952 unique categories. yuck\n",
+    "\n",
+    "sample = df.sample(frac = 0.05)\n"
```

Comment:
How many data entries did you end up with? If you have less than 10,000, that could explain why the trained model has such high precision. Make sure that there's no overlap between the head and the tail.